### PR TITLE
Perform repeat logic in compile phase instead.

### PR DESCRIPTION
**Pull request by josephschmitt**
This allows us to provide support for recursive rendering of items and categories. In order to support this, we're making a few changes to the API:
- The app maker no longer provides their own ng-repeat directive. We do this internally now. We provide access to a "suggestion" on the scope of a suggestion item which gives them access to their data.
- When rendering categories, we're now requiring a specific data format. Any sub-children of a category must be in a `children` array, and the data passed to the source function itself must also be an array of objects.

I've also stopped using lodash in the example to make @onlywei happy. Doing the categorizing loops by hand instead.

### CODE REVIEW

File: examples/category-results-materialize.html [hunk 1]
```diff
@@ -34,7 +37,6 @@
 
     <script src="../dist/ngc-omnibox.js"></script>
     <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/2.5.0/fuse.min.js"></script>
-    <script src="https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.2/lodash.min.js"></script>
```
Comment (onlywei) [inline examples/category-results-materialize.html]: sweet, no more lodash!

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js [hunk 2]
```diff
@@ -1,14 +1,53 @@
...
+export default function ngcOmniboxSuggestionsDirective($templateCache) {
   return {
     restrict: 'E',
     require: '^^ngcOmnibox',
     scope: true,
     controller: NgcOmniboxSuggestionsController,
     controllerAs: 'suggestions',
     compile(tElement) {
-      tElement.attr('role', 'listbox');
+      const el = tElement[0];
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: If you're never going to use the jqLite wrapped `tElement`, maybe consider using a destructured array as the arg?

```
compile([tElement]) {
  // now "tElement" is just a pure HTMLElement
}
```

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js [hunk 3]
```diff
@@ -1,14 +1,53 @@
...
+      el.setAttribute('role', 'listbox');
+
+      const categoryEl = el.querySelector(
+        'ngc-omnibox-suggestion-category, [ngc-omnibox-suggestion-category]'
+      );
+      const itemEl = el.querySelector(
+        'ngc-omnibox-suggestion-item, [ngc-omnibox-suggestion-item]'
+      );
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: hm, should we throw an error if we detect more than one `ngc-omnibox-suggestion-category` or `item`?

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js [hunk 4]
```diff
@@ -1,14 +1,53 @@
...
+
+      if (categoryEl) {
+        const categoryContainer = angular.element('<div></div>')[0];
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Why not just `document.createElement('div')` ?

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js [hunk 5]
```diff
@@ -1,14 +1,53 @@
...
+        categoryContainer.setAttribute('ng-repeat',
+            'suggestion in suggestion.children || omnibox.suggestions');
+        categoryEl.setAttribute('ng-if', 'suggestion.children');
+
+        const itemChildrenEl = angular.element('<div></div>')[0];
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Why not just `document.createElement('div')` ?

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js [hunk 6]
```diff
@@ -1,14 +1,53 @@
...
+        itemChildrenEl.setAttribute('ngc-omnibox-suggestion-item', '');
+        itemChildrenEl.setAttribute('ng-repeat', 'suggestion in suggestion.children');
+        itemChildrenEl.setAttribute('ng-include', '\'category-tmpl\'');
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Need to generate a unique identifier for the thing we put inside `$templateCache`, can't just use a static string.

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js [hunk 7]
```diff
@@ -1,14 +1,53 @@
...
+            itemChildrenEl.setAttribute(attr.name, attr.value);
+          }
+        }
+        itemEl.parentNode.appendChild(itemChildrenEl);
+
+        categoryContainer.appendChild(categoryEl);
+        categoryContainer.appendChild(itemEl);
+        itemEl.setAttribute('ng-if', '!suggestion.children');
+
+        $templateCache.put('category-tmpl', categoryContainer.innerHTML);
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Need to generate a unique identifier for the thing we put inside $templateCache, can't just use a static string.

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js [hunk 8]
```diff
@@ -1,14 +1,53 @@
...
+        if (itemEl.hasAttributes()) {
+          for (let i = 0; i < itemEl.attributes.length; i++) {
+            const attr = itemEl.attributes[i];
+            itemChildrenEl.setAttribute(attr.name, attr.value);
+          }
+        }
+        itemEl.parentNode.appendChild(itemChildrenEl);
+
+        categoryContainer.appendChild(categoryEl);
+        categoryContainer.appendChild(itemEl);
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: You've built a DOM structure in this code, I think it would help if there were a comment that contained the HTML representation of the DOM structure that is being built by this JavaScript. Of course, we would need to keep the HTML in comment 100% strictly up to date with the JavaScript because obsolete comments are worse than no comments at all...

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js (hunk 2 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: >Invalid attempt to destructure non-iterable instance

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js (hunk 7 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Done.

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js (hunk 8 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Done.

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js (hunk 6 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Done.

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js (hunk 5 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Done.

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js (hunk 4 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Done.

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js (hunk 3 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Yup, sorry, meant to do that originally. That's why I made it an `else if`

Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Oh wait nvm, I misread your comment. I don't think we throw an error, it'll just grab the first one and ignore the rest. I feel like we just say it'll only take one in the documentation. What do you think?

File: examples/category-results-materialize.html (hunk 1 above)
Comment (josephschmitt) [inline examples/category-results-materialize.html]: Thought you'd like that

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js [hunk 9]
```diff
@@ -1,14 +1,95 @@
...
+      itemChildrenEl.setAttribute('ngc-omnibox-suggestion-item', '');
+      itemChildrenEl.setAttribute('ng-repeat', 'suggestion in suggestion.children');
+      itemChildrenEl.setAttribute('ng-include', `'${templateCacheName}'`);
+      if (itemEl.hasAttributes()) {
+        for (let i = 0; i < itemEl.attributes.length; i++) {
+          const attr = itemEl.attributes[i];
+          itemChildrenEl.setAttribute(attr.name, attr.value);
+        }
+      }
+      itemChildrenEl.removeAttribute('ngc-omnibox-suggestion-item');
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Why are we adding this attribute on line 57 and then removing it here?

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js [hunk 10]
```diff
@@ -1,14 +1,95 @@
...
+   * the data structure of the Omnibox. The template will be different depending on whether it finds
+   * an ngc-omnibox-suggestion-category or not.
+   *
+   * Markup with an ngc-omnibox-suggestion-category:
+   *   <div ng-repeat="suggestion in suggestion.children || omnibox.suggestions">
+   *     <dl ngc-omnibox-suggestion-category ng-if="suggestion.children">
+   *       <dt>
+   *         <h5>{{suggestion.sample_category_title}}</h5>
+   *       </dt>
+   *       <dd>
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: I don't see any `<dd>` elements being created in the actual function. This HTML looks more like app-maker provided HTML rather than the structure that the function that's being documented is creating. This confuses me. Can we simplify to **just** what the function creates, with some pseudo-code if necessary? Like:

```
<category-container>
  <category-el>
  <item-children-el>
</category-container>
```

Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Hmm, that's tough since we're just modifying submitted markup vs creating much new markup of our own. Maybe I should document input -> output instead of just output.

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js (hunk 9 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: Done

File: spec/tests/angularComponent/ngcOmniboxControllerSpec.js [hunk 11]
```diff
@@ -31,27 +31,26 @@ describe('ngcOmnibox.angularComponent.ngcOmniboxController', () => {
 
       omniboxController.suggestions = 100;
       expect(omniboxController.hasSuggestions()).toBe(false);
-    });
 
-    it('should return false for empty Arrays and Objects', () => {
-      omniboxController.suggestions = [];
+      omniboxController.suggestions = {test: 'me'};
       expect(omniboxController.hasSuggestions()).toBe(false);
```
Comment (onlywei) [inline spec/tests/angularComponent/ngcOmniboxControllerSpec.js]: Should it error instead of being false? I feel like at some point there should be an error.

File: src/angularComponent/ngcOmniboxSuggestionItemController.js [hunk 12]
```diff
@@ -1,3 +1,16 @@
 export default class NgcOmniboxSuggestionItemController {
+  static get $inject() {
+    return ['$scope'];
+  }
+  constructor($scope) {
+    this.$scope = $scope;
+  }
+  $onInit() {
+    this.omnibox = this.$scope.omnibox;
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionItemController.js]: Don't you get this for free using `require`?

File: src/angularComponent/ngcOmniboxSuggestionItemDirective.js [hunk 13]
```diff
@@ -1,23 +1,21 @@
-import {SUGGESTION_ITEM_NAME} from './ngcOmniboxController.js';
 import NgcOmniboxSuggestionItemController from './ngcOmniboxSuggestionItemController.js';
 
-export default function ngcOmniboxSuggestionsDirective() {
+export default function ngcOmniboxSuggestionItemDirective() {
   return {
     restrict: 'AE',
     require: ['^^ngcOmnibox', '^^ngcOmniboxSuggestions'],
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionItemDirective.js]: Switch to object syntax `require` instead of array syntax to make my previous comment work.

File: src/angularComponent/ngcOmniboxSuggestionItemDirective.js [hunk 14]
```diff
@@ -1,23 +1,21 @@
...
     compile(tElement) {
-      // Allows us to use document.getElementsByName which is fast AND returns a live-updating
-      // HTMLCollection
-      tElement.attr('name', SUGGESTION_ITEM_NAME);
       tElement.attr('role', 'option');
 
       return {
         pre(scope, iElement, iAttrs, [omnibox]) {
           scope.omnibox = omnibox;
-          scope.suggestionItem.getIndex = () => omnibox.getSuggestionItemIndex(iElement[0]);
```
Comment (onlywei) [inline src/angularComponent/ngcOmniboxSuggestionItemDirective.js]: sweet! I like this!

File: spec/tests/angularComponent/ngcOmniboxControllerSpec.js (hunk 11 above)
Comment (josephschmitt) [inline spec/tests/angularComponent/ngcOmniboxControllerSpec.js]: Hmm, my initial inclination is no. A function that starts with `is` or `has` should return a `Boolean` imo, not error out.

File: src/angularComponent/ngcOmniboxSuggestionItemController.js (hunk 12 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionItemController.js]: Done

File: src/angularComponent/ngcOmniboxSuggestionItemDirective.js (hunk 13 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionItemDirective.js]: Done. Didn't know I could use that syntax with directives.

File: src/angularComponent/ngcOmniboxSuggestionItemDirective.js (hunk 14 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionItemDirective.js]: Yea that's all gone now. Much simpler.

File: src/angularComponent/ngcOmniboxSuggestionsDirective.js (hunk 10 above)
Comment (josephschmitt) [inline src/angularComponent/ngcOmniboxSuggestionsDirective.js]: I ended up adding the tests, but I want this to be in its own PR. This one's getting beefy enough.
